src/raftstore/store/snap.rs

```diff
@@ -617,6 +618,750 @@ mod v1 {
+            }
+            self.cf_index = index;
+            Ok(())
+        }
+
+        fn add_kv(&mut self, k: &[u8], v: &[u8]) -> io::Result<()> {
+            let mut cf_file = &mut self.cf_files[self.cf_index];
+            let mut writer = cf_file.sst_writer.as_mut().unwrap();
+            if let Err(e) = writer.add(k, v) {
+                return Err(io::Error::new(ErrorKind::Other, e));
+            }
+            cf_file.kv_count += 1;
+            Ok(())
+        }
+
+        fn save_cf_files(&mut self) -> io::Result<()> {
+            let mut total_size = 0;
+            for cf_file in &mut self.cf_files {
+                if cf_file.kv_count == 0 {
+                    try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                    cf_file.size = 0;
+                } else {
+                    let writer = cf_file.sst_writer.as_mut().unwrap();
+                    if let Err(e) = writer.finish() {
+                        return Err(io::Error::new(ErrorKind::Other, e));
+                    }
+                    try!(fs::rename(&cf_file.tmp_path, &cf_file.path));
+                    let size = try!(get_file_size(&cf_file.path));
+                    cf_file.size = size;
+                    total_size += size;
+                }
+                cf_file.digest = try!(calc_checksum(&cf_file.path));
+            }
+            let mut size_track = self.size_track.wl();
+            *size_track = size_track.saturating_add(total_size);
+            Ok(())
+        }
+
+        fn save_meta_file(&mut self) -> raft::Result<()> {
+            let mut v = vec![];
+            box_try!(self.meta_file.data.write_to_vec(&mut v));
+            let mut f = self.meta_file.file.as_mut().unwrap();
+            try!(f.write_all(&v[..]));
+            try!(f.flush());
+            try!(fs::rename(&self.meta_file.tmp_path, &self.meta_file.path));
+            Ok(())
+        }
+
+        fn do_build(&mut self, snap: &DbSnapshot, region: &Region) -> raft::Result<()> {
+            if self.exists() {
+                match self.validate() {
+                    Ok(()) => return Ok(()),
+                    Err(e) => {
+                        error!("[region {}] file {} is invalid, will rebuild: {:?}",
+                               region.get_id(),
+                               self.path(),
+                               e);
+                        try!(self.try_delete());
+                        try!(self.init_for_building());
+                    }
+                }
+            }
+
+            let t = Instant::now();
+            let mut snap_size = 0;
+            let mut snap_key_count = 0;
+            let (begin_key, end_key) = (enc_start_key(region), enc_end_key(region));
+            for cf in snap.cf_names() {
+                if !need_to_pack(cf) {
+                    continue;
+                }
+                try!(self.next_file(cf.to_owned()));
```

Comment:
seem that it is not for next file, but to get the correct writer index with the CF.